src/binary/merkle_tree.rs

```diff
@@ -239,41 +255,49 @@ where
             .iter()
             .unzip();
 
-        let mut current_head = None;
-        let peaks = &peaks.as_slice()[1..]; // Omit the root.
-        for peak in peaks.iter() {
-            let key = peak.in_order_index();
-            let node = self
-                .storage
-                .get(&key)?
-                .ok_or(MerkleTreeError::LoadError(key))?
-                .into_owned()
-                .into();
-            let next = Box::new(Subtree::<Node>::new(node, current_head));
-            current_head = Some(next);
-        }
-
-        self.head = current_head;
+        peaks
+    }
 
-        Ok(())
+    fn root_position(&self) -> Position {
+        self.peak_positions()[0]
     }
 
-    fn root_node(&mut self) -> Result<Option<Node>, StorageError> {
-        let root_node = match self.head {
-            None => None,
-            Some(ref initial) => {
-                let mut current = initial.clone();
-                while current.next().is_some() {
-                    let mut head = current;
-                    let mut head_next = head.take_next().unwrap();
-                    current = self.join_subtrees(&mut head_next, &mut head)?
-                }
-                Some(current.node().clone())
-            }
-        };
+    fn root_node(
+        &self,
+        scratch_storage: &mut StorageMap<NodesTable>,
+    ) -> Result<Option<Node>, StorageError> {
+        let root_node = self
+            .head
+            .as_ref()
+            .map(|head| build_root_node(head, scratch_storage))
+            .transpose()
+            .unwrap(); // Safety: scratch_storage is infallible
```

Comment:
Could you use `expect` here then with a proper message, please?=)

And seems we can remove `Result` from the method's signature.